In [3]:
from dbfread import DBF
import pandas as pd
from pandas import DataFrame

## Load 10 years of accident data, from 2007 to 2016

In [4]:
dbf07= DBF('accident/accident2007.dbf')
dbf08= DBF('accident/accident2008.dbf')
dbf09= DBF('accident/accident2009.dbf')
dbf10= DBF('accident/accident2010.dbf')
dbf11 = DBF('accident/accident2011.dbf')
dbf12 = DBF('accident/accident2012.dbf')
dbf13 = DBF('accident/accident2013.dbf')
dbf14 = DBF('accident/accident2014.dbf')
dbf15 = DBF('accident/accident2015.dbf')
dbf16 = DBF('accident/accident2016.dbf')
accidents07 = DataFrame(iter(dbf07))
accidents08 = DataFrame(iter(dbf08))
accidents09 = DataFrame(iter(dbf09))
accidents10 = DataFrame(iter(dbf10))
accidents11 = DataFrame(iter(dbf11))
accidents12 = DataFrame(iter(dbf12))
accidents13 = DataFrame(iter(dbf13))
accidents14 = DataFrame(iter(dbf14))
accidents15 = DataFrame(iter(dbf15))
accidents16 = DataFrame(iter(dbf16))


First, we want to combine accidents10 ~ accidents16 to one dataframe. Since not all of the accident data downloaded from the U.S. Department of Transportation have the same features, by using the `jion:inner` option in `pd.concat` function, we can get the intersection of features.

In [5]:
# rename column name in frame07 so that columns names are the same with other frames
accidents07.rename(columns={'latitude': 'LATITUDE', 'longitud': 'LONGITUD'}, inplace=True)

In [6]:
# take a look inside how the accident data file looks like
allaccidents = pd.concat([accidents07,accidents08,accidents09,accidents10,accidents11,accidents12,accidents13,accidents14,accidents15,accidents16], axis=0,join='inner')
pd.set_option('display.max_columns', 100)
allaccidents.head()

STATE  COUNTY  MONTH   DAY  HOUR  MINUTE  VE_TOTAL  PERSONS  PEDS  NHS  \
0    1.0    73.0    1.0   2.0  23.0    15.0       1.0      1.0   0.0  0.0   
1    1.0    19.0    1.0  30.0  13.0     5.0       3.0      4.0   0.0  0.0   
2    1.0     9.0    1.0  14.0  12.0    13.0       2.0      5.0   0.0  0.0   
3    1.0    15.0    1.0   1.0  15.0     0.0       2.0      2.0   0.0  1.0   
4    1.0    51.0    1.0   6.0  19.0    35.0       2.0      5.0   0.0  0.0   

   ROUTE  SP_JUR  HARM_EV  MAN_COLL  REL_ROAD  LGT_COND  WEATHER1  WEATHER2  \
0    4.0     0.0     42.0       0.0       4.0       2.0       1.0       0.0   
1    3.0     0.0     12.0       5.0       1.0       1.0       1.0       0.0   
2    2.0     0.0     12.0       7.0       1.0       1.0       1.0       0.0   
3    2.0     0.0     12.0       4.0       1.0       1.0       1.0       0.0   
4    3.0     0.0     12.0       5.0       1.0       3.0       1.0       0.0   

   NOT_HOUR  NOT_MIN  ARR_HOUR  ARR_MIN  HOSP_HR  HOSP_MN  SCH_BUS  CF1  CF2  \
0      23.0     16.0      23.0     20.0      0.0      0.0      0.0  0.0  0.0   
1      13.0     10.0      13.0     11.0     99.0     99.0      0.0  0.0  0.0   
2      12.0     16.0      12.0     18.0     99.0     99.0      0.0  0.0  0.0   
3      99.0     99.0      15.0      5.0      0.0      0.0      0.0  0.0  0.0   
4      19.0     35.0      19.0     41.0     99.0     99.0      0.0  0.0  0.0   

   CF3  FATALS  DAY_WEEK  DRUNK_DR  ST_CASE    CITY  MILEPT    YEAR  TWAY_ID  \
0  0.0     1.0       3.0       0.0  10001.0     0.0     0.0  2007.0     1493   
1  0.0     2.0       3.0       0.0  10002.0   630.0     0.0  2007.0    SR-68   
2  0.0     1.0       1.0       0.0  10003.0  2330.0     0.0  2007.0  US-SR53   
3  0.0     1.0       2.0       0.0  10004.0     0.0  2468.0  2007.0   US-SR1   
4  0.0     1.0       7.0       0.0  10005.0  2063.0     0.0  2007.0    SR-14   

   TWAY_ID2     RAIL  VE_FORMS  WEATHER   LATITUDE   LONGITUD  
0  00000000  0000000       1.0      1.0  33.455839 -87.017928  
1       270  0000000       3.0      1.0  34.160597 -85.678075  
2  00000000  0000000       2.0      1.0  33.975717 -86.496347  
3      7260  0000000       2.0      1.0  33.853258 -85.915622  
4      I-65  0000000       2.0      1.0  32.488833 -86.408936

In [7]:
# allaccidents['ST_CASE'] = 'n' + df['ST_CASE']

In [ ]:
allaccidents.shape

(320874, 42)

The allaccidents table recorded 320874 accidents from 2010-2016, and it has 42 features. Here are the meaning of some of the features according to the `FARS Analytical User’s Manual`.

### Explaination of variables
*VE_TOTAL*: Number of Vehicle in crash <br/>
*VE_FORMS*: Number of Motor Vehicles in Transport (MVIT) <br/>
*PED*: Number of Persons Not in Motor Vehicles <br/>
*NHS*: National Highway System<br/>
*ROUTE*: Route Signing <br/>
*SP_JUR*: Special Jurisdiction <br/>
*HARM_EV*: First Harmful Event<br/>
*TWAY_ID , TWAY_ID2* : Trafficway Identifier <br/>
*MILEPT*: Milepoint <br/>
*SP_JUR*: Special Jurisdiction<br/>
*HARM_EV*: injury or damage producing First Harmful Event <br/> 
*MAN_COLL*:Manner of Collision  <br/> 
*RELJCT1, RELJCT2*: Relation to Junction- Within Interchange Area, Specific Location. <br/>
*TYP_INT*: Type of Intersection <br/>
*REL_ROAD*: Relation to Trafficway <br/>
*LGT_COND*: Light Condition<br/> 
*NOT_HOUR,MIN*: Min, Hour of Notification <br/>
*ARR_HOUR,MIN*: Hour, Min arrival at scene <br/>
*HOSP_HR,MIN*: Hour, Min arrival at hospital <br/>
*CF1, CF2, CF3*:Related Factors- Crash Level, factors related to the crash <br/>
*FATALS*: Fatalities<br/>
*DRUNK_DR*: Number of Drinking Drivers<br/> 
*RAIL*: Rail Grade Crossing Identifier<br/>

For more detailed information, please refer to `FARS Analytical User’s Manual`.

## Select variables and rename variables

Observed from the table above, some of the variables in the table are not very readable. Therefore, in order to make it easier to understand the variables,we renamed some of the variables according to `FARS Analytical User’s Manual`  downloaded from the  `U.S. Department of Transportation`  website.

### Load vehicle data file which contains mortality rate

We also want to study the mortality rate of fatal accidents. The data element “Fatalities in Vehicle” in the Vehicle data file from the `U.S. Department of Transportation` website provides the number of deaths in a vehicle.

In [ ]:
vdbf07= DBF('vehicle_deaths/vehicle2007.dbf')
vdbf08= DBF('vehicle_deaths/vehicle2008.dbf')
vdbf09= DBF('vehicle_deaths/vehicle2009.dbf')
vdbf10= DBF('vehicle_deaths/vehicle2010.dbf')
vdbf11= DBF('vehicle_deaths/vehicle2011.dbf')
vdbf12= DBF('vehicle_deaths/vehicle2012.dbf')
vdbf13= DBF('vehicle_deaths/vehicle2013.dbf')
vdbf14= DBF('vehicle_deaths/vehicle2014.dbf')
# vdbf15= DBF('vehicle_deaths/vehicle2015.csv')
vdbf16= DBF('vehicle_deaths/vehicle2016.dbf')
vehicle07 = DataFrame(iter(vdbf07))
vehicle08 = DataFrame(iter(vdbf08))
vehicle09 = DataFrame(iter(vdbf09))
vehicle10 = DataFrame(iter(vdbf10))
vehicle11 = DataFrame(iter(vdbf11))
vehicle12 = DataFrame(iter(vdbf12))
vehicle13 = DataFrame(iter(vdbf13))
vehicle14 = DataFrame(iter(vdbf14))
# vehicle15 = pd.read_csv('vehicle_deaths/vehicle2015.csv')
vehicle16 = DataFrame(iter(vdbf16))

In [ ]:
vehicle07['YEAR']='2007.0'
vehicle08['YEAR']='2008.0'
vehicle09['YEAR']='2009.0'
vehicle10['YEAR']='2010.0'
vehicle11['YEAR']='2011.0'
vehicle12['YEAR']='2012.0'
vehicle13['YEAR']='2013.0'
vehicle14['YEAR']='2014.0'
# vehicle15['YEAR']='2015.0'
vehicle16['YEAR']='2016.0'

In [ ]:
accidents16['DEATHS']=list(vehicle16.groupby('ST_CASE')['DEATHS'].sum())#ADD DEATHS NUMBER TO EACH ACCIDENT
#accidents15['DEATHS']=list(vehicle16.groupby('ST_CASE')['DEATHS'].sum())
accidents14['DEATHS']=list(vehicle14.groupby('ST_CASE')['DEATHS'].sum())
accidents13['DEATHS']=list(vehicle13.groupby('ST_CASE')['DEATHS'].sum())
accidents12['DEATHS']=list(vehicle12.groupby('ST_CASE')['DEATHS'].sum())
accidents11['DEATHS']=list(vehicle11.groupby('ST_CASE')['DEATHS'].sum())
accidents10['DEATHS']=list(vehicle10.groupby('ST_CASE')['DEATHS'].sum())
accidents09['DEATHS']=list(vehicle09.groupby('ST_CASE')['DEATHS'].sum())
accidents08['DEATHS']=list(vehicle08.groupby('ST_CASE')['DEATHS'].sum())
accidents07['DEATHS']=list(vehicle07.groupby('ST_CASE')['DEATHS'].sum())

In [ ]:
#combine all vehicle file
allaccidents = accidents07.append([accidents08,accidents09,accidents10,accidents11,accidents12,accidents13,accidents14,accidents15,accidents16])
accidents = allaccidents[['YEAR','ST_CASE','STATE','VE_TOTAL','PERSONS','FATALS','DEATHS','MONTH','DAY_WEEK','HOUR','NHS','LATITUDE','LONGITUD','MAN_COLL','LGT_COND','WEATHER','ARR_HOUR','ARR_MIN','CF1','DRUNK_DR']]
accidents.rename(columns={'ST_CASE':'CASE_NUM','VE_TOTAL':'NUM_VEHICLE','NHS': 'HIGHWAY', 'MAN_COLL': 'COLLISION_TYPE','LGT_COND':'LIGHT_CONDITION','CF1':'CRASH_FACTOR','DRUNK_DR':'DRUNK_DRIVE'}, inplace=True)
accidents['MONTH'] = accidents['MONTH'].map({1.0:'January', 2.0:'February', 3.0: 'March', 4.0:'April', 5.0:'May', 6.0:'June', 7.0:'July', 8.0:'August',9.0: 'September', 10.0:'October', 11.0:'November', 12.0:'December'})
accidents['DAY_WEEK']= accidents['DAY_WEEK'].map({1.0:'Sunday',2.0:'Monday', 3.0:'Tuesday', 4.0: 'Wednesday', 5.0:'Thursday', 6.0:'Friday', 7.0:'Saturday'})
accidents['HIGHWAY'] = accidents['HIGHWAY'].map({1.0:'On',0.0:'Off',9.0:'Unknow'})
accidents['COLLISION_TYPE'] = accidents['COLLISION_TYPE'].map({0.0:'Not Collision',1.0:'Rear-End',2.0:'Head-On',3.0:'Rear-to-Rear',4.0:'Angle',5.0:'Sideswipe, Same Direction',6.0:'Sideswipe, Opposite Direction',7.0:'Sideswipe, Unknown Direction',9.0:'Unknown'})
accidents['LIGHT_CONDITION'] = accidents['LIGHT_CONDITION'].map({1.0:'Daylight',2.0:'Dark' ,3.0:'Dark',5.0:'Dusk',6.0:'Dark',4.0:'Dawn', 7.0:'Other',8.0 :'Not Report', 9.0:'Not Report'})
# accidents['WEATHER'] = accidents['WEATHER'].map({0.0:'Normal',1.0:'Clear',2.0:'Rain',3.0
accidents.head()